### PR TITLE
parachain-system: update RelevantMessagingState according to the unincluded segment

### LINKED ISSUE #2471: Ensure outgoing XCMP respects unincluded segment
@rphmeier 
```rust
/// A means of figuring out what outbound XCMP messages should be being sent.
pub trait XcmpMessageSource {
	/// Take a single XCMP message from the queue for the given `dest`, if one exists.
	fn take_outbound_messages(maximum_channels: usize) -> Vec<(ParaId, Vec<u8>)>;
}
```
I believe this trait not only should account for max messages taken from the storage, but also for segment limitations if we want to make this panic correct.
Probably primitives for segment should be extracted into common crate to be accessible from both pallets.

_Originally posted by @slumber in https://github.com/paritytech/cumulus/pull/2438#discussion_r1172857195_

### PR DESCRIPTION
**Pull request by rphmeier**
Closes #2471  ,#2950

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3279072

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3287333

File: pallets/parachain-system/src/lib.rs [hunk 1]
```diff
@@ -1227,6 +1240,46 @@ impl<T: Config> Pallet<T> {
...
+
+				let c = &mut channels[i].1;
+
+				c.total_size = (c.total_size + used.total_bytes).max(c.max_total_size);
+				c.msg_count = (c.msg_count + used.msg_count).max(c.max_capacity);
+			}
+
+			let upward_capacity = &mut messaging_state.relay_dispatch_queue_remaining_capacity;
+			upward_capacity.remaining_count =
+				upward_capacity.remaining_count.saturating_sub(used_bandwidth.ump_msg_count);
```
Comment (BradleyOlson64) [inline pallets/parachain-system/src/lib.rs:1277]: Sanity checking, do we want to be subtracting the entire `used_bandwidth` of the segment here each time?

Comment (rphmeier) [reply pallets/parachain-system/src/lib.rs:1277]: Yeah. What we read from the relay-chain state is the total capacity as-of the most recently included block. We then subtract the used bandwidth in the unincluded segment from it.

**Review by BradleyOlson64 — APPROVED** [review]

**Review by slumber — APPROVED** [review]
Looks correct, although I'd say tests are quite important to have.

File: pallets/parachain-system/src/lib.rs [hunk 2]
```diff
@@ -252,22 +256,28 @@ pub mod pallet {
...
+					Some(limits) => (
+						limits.relay_dispatch_queue_remaining_capacity.remaining_count,
+						limits.relay_dispatch_queue_remaining_capacity.remaining_size,
+					),
+					None => {
+						debug_assert!(
+							false,
+							"relevant messaging state is promised to be set until `on_finalize`; \
+								qed",
+						);
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply pallets/parachain-system/src/lib.rs:274]: this pattern was already present in the code, so I just chose to preserve that.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3300244

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3300928

File: pallets/parachain-system/src/tests.rs [hunk 3]
```diff
@@ -137,15 +137,23 @@ fn send_message(dest: ParaId, message: Vec<u8>) {
...
-		let mut taken = 0;
+		let mut taken_messages = 0;
+		let mut taken_bytes = 0;
 		let mut result = Vec::new();
 		SENT_MESSAGES.with(|ms| {
 			ms.borrow_mut().retain(|m| {
 				let status = <Pallet<Test> as GetChannelInfo>::get_channel_status(m.0);
-				let ready = matches!(status, ChannelStatus::Ready(..));
-				if ready && !ids.contains(&m.0) && taken < maximum_channels {
+				let ChannelStatus::Ready(max_size_now, max_size_ever) = status else { return false };
```
Comment (rphmeier) [inline pallets/parachain-system/src/tests.rs:146]: This change is wrong. As this is within a `retain`, returning false (e.g. when the channel is full) would prune the message forever.

Comment (slumber) [reply pallets/parachain-system/src/tests.rs]: yes, my bad, pushed a fix

Comment (rphmeier) [reply pallets/parachain-system/src/tests.rs]: I pushed a more comprehensive version + added a bunch more to the test you wrote, just while I was playing around with it.

**Comment by rphmeier** [comment]
bot merge

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3303140

**Comment by slumber** [comment]
bot merge

**Pull request merged**